### PR TITLE
Implemented caching mechanism for API store

### PR DESCRIPTION
**Pull request by gustavosbarreto**
Implemented caching mechanism to reduce database read/write operations from API store.

### Dependencies

- [x] #888 api: Remove dependency on bson package from API services

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -31,6 +31,10 @@ SHELLHUB_REDIRECT_TO_HTTPS=false
 # Values: a valid domain name
 SHELLHUB_DOMAIN=localhost
 
+# Enable store cache (improve API response time)
+# NOTICE: Redis is required
+SHELLHUB_ENABLE_STORE_CACHE=false
```
Comment (otavio) [inline .env:36]: `STORE_CACHE`.

File: agent/main.go [hunk 2]
```diff
@@ -164,7 +164,7 @@ func main() {
 		}()
 	}
 
-	ticker := time.NewTicker(time.Duration(opts.KeepAliveInterval) * time.Second)
+	ticker := time.NewTicker(time.Duration(10 * time.Second))
```
Comment (otavio) [inline agent/main.go:167]: Is this intended?

File: api/main.go [hunk 3]
```diff
@@ -9,13 +9,15 @@ import (
...
+	storecache "github.com/shellhub-io/shellhub/api/store/cache"
 	"github.com/shellhub-io/shellhub/api/store/mongo"
 	mgo "go.mongodb.org/mongo-driver/mongo"
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
 type config struct {
-	MongoUri string `envconfig:"mongo_uri" default:"mongodb://mongo:27017"`
+	MongoUri         string `envconfig:"mongo_uri" default:"mongodb://mongo:27017"`
+	EnableStoreCache bool   `envconfig:"enable_store_cache" default:false"`
```
Comment (otavio) [inline api/main.go:20]: StoreCache

**Review by otavio — CHANGES_REQUESTED** [review]

File: agent/main.go (hunk 2 above)
Comment (gustavosbarreto) [reply agent/main.go:167]: WIP

Comment (otavio) [reply agent/main.go:167]: I'll keep this open so we don't forget to solve this prior to merging it.

File: api/store/mongo/device_store.go [hunk 4]
```diff
@@ -308,10 +318,23 @@ func (s *Store) DeviceGetByName(ctx context.Context, name, tenant string) (*mode
 }
 
 func (s *Store) DeviceGetByUID(ctx context.Context, uid models.UID, tenant string) (*models.Device, error) {
-	device := new(models.Device)
+	var device *models.Device
+
+	if err := s.cache.Get(ctx, strings.Join([]string{"device", string(uid)}, "/"), &device); err != nil {
+		fmt.Println(err)
```
Comment (otavio) [inline api/store/mongo/device_store.go:324]: Shouldn't here we return?

File: api/store/mongo/device_store.go [hunk 5]
```diff
@@ -308,10 +318,23 @@ func (s *Store) DeviceGetByName(ctx context.Context, name, tenant string) (*mode
...
+	if device != nil {
+		return device, nil
+	}
+
 	if err := s.db.Collection("devices").FindOne(ctx, bson.M{"tenant_id": tenant, "uid": uid}).Decode(&device); err != nil {
 		return nil, err
 	}
 
+	if err := s.cache.Set(ctx, strings.Join([]string{"device", string(uid)}, "/"), device, time.Minute); err != nil {
+		fmt.Println(err)
```
Comment (otavio) [inline api/store/mongo/device_store.go:336]: ... and here?

**Review by otavio — CHANGES_REQUESTED** [review]

File: api/store/mongo/device_store.go (hunk 4 above)
Comment (gustavosbarreto) [reply api/store/mongo/device_store.go:324]: WIP
https://github.com/shellhub-io/shellhub/blob/94d3a9b492f4599a8dfd95b4668dcd47ccd87b13/api/store/cache/cache_redis.go#L30

File: api/store/mongo/device_store.go (hunk 5 above)
Comment (gustavosbarreto) [reply api/store/mongo/device_store.go:336]: WIP
https://github.com/shellhub-io/shellhub/blob/94d3a9b492f4599a8dfd95b4668dcd47ccd87b13/api/store/cache/cache_redis.go#L30

File: api/store/mongo/device_store.go (hunk 4 above)
Comment (otavio) [reply api/store/mongo/device_store.go:324]: No really. If `Get` handle the cache failure (and it seems it does) than errors getting here are indeed errors.

Comment (gustavosbarreto) [reply api/store/mongo/device_store.go:324]: An error in the cache should not result in application failure.

Comment (gustavosbarreto) [reply api/store/mongo/device_store.go:324]: As it is in WIP, a print is enough, when finished it will become just a warning message.

**Review by otavio — CHANGES_REQUESTED** [review]

File: api/store/cache/cache_redis.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+// Get gets the cache value for the given key
+// NOTE: missing key is not an error
+func (c *redisCache) Get(ctx context.Context, key string, value interface{}) error {
+	err := c.cache.Get(ctx, key, value)
+	if err == rediscache.ErrCacheMiss {
+		return nil
+	}
+
+	return err
```
[the remark this answers is not in the document]
Comment (gustavosbarreto) [reply api/store/cache/cache_redis.go:38]: // Get gets the cache value for the given key
// NOTE: missing key is not an error

File: api/store/mongo/device_store.go [hunk 7]
```diff
@@ -183,6 +184,10 @@ func (s *Store) DeviceDelete(ctx context.Context, uid models.UID) error {
 		return err
 	}
 
+	if err := s.cache.Delete(ctx, string(uid)); err != nil {
+		logrus.Warning(err)
```
[the remark this answers is not in the document]
Comment (gustavosbarreto) [reply api/store/mongo/device_store.go:190]: An error in the cache should not result in application failure.

File: api/store/cache/cache_redis.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (otavio) [reply api/store/cache/cache_redis.go:38]: I saw this; what I am saying is it ought to be.

File: api/store/mongo/device_store.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (otavio) [reply api/store/mongo/device_store.go:190]: Sure not but it should be logged accordingly.

**Pull request merged**